pkg/kubernetes/scripts/virtual-machines/index.jsx

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import 'regenerator-runtime/runtime' // required for library initialization
+import React from 'react'
+import { createStore, applyMiddleware, compose } from 'redux'
+import createSagaMiddleware from 'redux-saga'
+import { Provider } from 'react-redux'
+
+import reducers from './reducers.jsx'
+import rootSaga from './sagas.jsx'
+import * as actionCreators from './action-creators.jsx'
+import VmsListing from './components/VmsListing.jsx'
+import { getValues } from './utils.jsx'
+
+const sagaMiddleware = createSagaMiddleware()
+let reduxStore
+
+function initReduxStore() {
+    const initialState = {
+        vms: []
+    }
+    const middleware = [ sagaMiddleware ]
+    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
+    const storeEnhancers = composeEnhancers(applyMiddleware(...middleware))
+    reduxStore = createStore(reducers, initialState, storeEnhancers)
+}
+
+function addKubeLoaderListener ($scope, kubeLoader, kubeSelect) {
+    // register load callback( callback, until )
+    kubeLoader.listen(function() {
+        const vms = kubeSelect().kind('VirtualMachine')
+        reduxStore.dispatch(actionCreators.setVms(getValues(vms)))
+    }, $scope);
+
+    // enable watching( watched-entity-type, until )
+    kubeLoader.watch('VirtualMachine', $scope);
+}
+
+const VmsPlugin = () => (
+    <Provider store={reduxStore} >
+        <VmsListing />
+    </Provider>
+)
+
+function addScopeVarsToStore ($scope) {
+    $scope.$watch(
+        (scope => scope.settings),
+        (newSettings => reduxStore.dispatch(actionCreators.setSettings(newSettings))))
+}
+
+/**
+ *
+ * @param {$rootScope.Scope} $scope 'VirtualMachinesCtrl' controller scope
+ * @param {kubeLoader} kubeLoader
+ * @param {kubeSelect} kubeSelect
+ */
+function init($scope, kubeLoader, kubeSelect) {
+    initReduxStore()
+    sagaMiddleware.run(rootSaga)
```

Comment:
Are you actually using sagas anywhere? At least to my ignorant eye it just looks like you have the skeleton here, but it doesn't do anything (yet?). I may sound like an old man, but I find it a bit hard to follow all that boilerplate in reducers, action-types, and action-creators (particularly as this page does not actually have any user-visible actions, it's just displaying state). It's mostly "your" code, so I don't object, but I'm at least trying to understand what's happening :-)

Reply:
Your sharp eye correctly recognized that there are no sagas yet. If it's a problem it can be removed from this pull. Commit 'kubernetes: Virtual Machines view - infra' is meant to set up complete react-redux-saga environment. Sagas will be used in future patches performing "write" operations.

I agree it's not exactly concise. However AFAIK it's the best of structuring state management for React.

Action is unfortunately overloaded expression here. In context of redux it's used to describe an event changing the state.

Reply:
It's okay if you are going to use them soon, I just wanted to confirm that it's not just some leftover boilerplate. Thanks! (Going to review the rest of the updated PR on Monday, thanks for the cleanups!)